### ISSUE TITLE
Failed to open stream: Permission denied

### ISSUE THREAD

**Issue by jaydrogers** [serversideup/docker-php#179]
### Affected Docker Images

All

### Current Behavior

There are many reports where this error randomly appears, usually on fresh deployments.

```
Failed to open stream: Permission denied
```

#### Examples
https://github.com/serversideup/docker-php/pull/172#issue-1654792968

**Bugflow**
<img width="1723" alt="image" src="https://github.com/serversideup/docker-php/assets/3174134/8c1bb3fe-6a75-4037-be87-5641768f064e">

### Expected Behavior

There is a discussion where we may be able to adjust S6 Overlay and have the commands run correctly (instead of root)

https://github.com/serversideup/docker-php/discussions/71#discussioncomment-5841649

### Steps To Reproduce

TBD

### Host Operating System

Linux

### Docker Version

```markdown
N/A
```

**Comment by jaydrogers** [comment]
Adding my error message in here too for future reference 


<img width="1949" alt="image" src="https://github.com/serversideup/docker-php/assets/3174134/9ecd7ddb-935b-4958-b13e-2624e1e4d309">

**Comment by danilopolani** [comment]
Wanted to add a few things on this issue.  
I copied all the project files with the proper user ownership like this:

```Dockerfile
COPY --chown=$PUID:$PGID ./composer.json ./composer.lock ./
RUN composer install --no-cache --no-dev --no-scripts --no-autoloader --ansi --no-interaction

COPY --chown=$PUID:$PGID . .
RUN composer dump-autoload -o
``` 

Now the issue is that it can't create cache files:

> Unable to create lockable file: /var/www/html/storage/framework/cache/data/  
  fa/f2/faf26108316c685830630fa2070ff0a7b212a06a. Please ensure you have perm  
  ission to create files in this location.

I've tried giving to the whole `storage` folder all the permissions and ownership to both `$PUID:$PGID` or `www-data:www-data`, but no luck so far. Also with the `beta` I got no luck. What's the user running the webserver?

**Comment by jaydrogers** [comment]
Thanks for chiming in. The user running PHP is `webuser`, so you might need to do a `su` before running your composer commands.

Also -- I am not sure why you would need two copy commands in there. Also, the default web root is `/var/www/html`. If you're running your build within the Docker build process, then it would look something like this:

```Dockerfile
COPY --chown=$PUID:$PGID ./composer.json ./composer.lock /var/www/html
RUN su webuser && \
    composer install --no-cache --no-dev --no-scripts --no-autoloader --ansi --no-interaction && \
    composer dump-autoload -o
```

**Comment by danilopolani** [comment]
Right, splitting the command is not needed. At the end I've cleared all my docker images and containers and now this is the full, working code, no errors related to `laravel.log` nor cache creation, if it may help someone else too.

```Dockerfile
FROM serversideup/php:8.2-fpm-nginx as base

ENV AUTORUN_LARAVEL_MIGRATION=true
ENV SSL_MODE=off
ENV APP_ENV=production

FROM base as production

COPY --chown=$PUID:$PGID . .

RUN composer install --no-cache --no-dev --no-scripts --no-autoloader --ansi --no-interaction \
    && composer dump-autoload -o
``` 

I had to create a separate Dockerfile because I added (not pasted here) some tweaks to opcache, more extensions and so on. Thank you for the support and the image :)

**Comment by tsterker** [comment]
@jaydrogers I'm currently trying to follow up on this topic and understand the issue mentioned [above](https://github.com/serversideup/docker-php/issues/179#issuecomment-1563218182) and how #180 could be the cause for this. 

ℹ️ **Is there an example that works with the current setup and breaks with #180?**

e.g. you mention a GitLab CI + DIND setup [above](https://github.com/serversideup/docker-php/issues/179#issuecomment-1563218182).

> GitLab CI + DIND (Docker-in-docker) uses root and the UID of 0. With the commands using UID 9999 by default, the CI machine cannot create a vendor directory on the first step of composer install

Is this related to building docker images? In this case, maybe [kaniko](https://docs.gitlab.com/ee/ci/docker/using_kaniko.html) could be an option to build images without privileged mode?

That said, if there is a general compatibility issue with the ubiquitous DIND it's still a problem that would need solving.

**Comment by jaydrogers** [comment]
### Just adding a note for myself...
S6 Overlay has some changes that were released on May 4th. I might want to explore these changes since they talk specifically about the `USER` directive: https://github.com/just-containers/s6-overlay#notes

**Comment by jaydrogers** [comment]
Just adding a note, https://github.com/serversideup/docker-php/pull/180 has been reverted.

## New approach
- I am going to explore if the `USER` directive in `v3.1.5.0` of S6 Overlay is a better approach

**Comment by elcapo** [comment]
I'm having the same issue here. Just for context, if I check the owner of the files using a disposable container by running `spin run php ls -la`, I get a list of files owned by 1000:1000:

~~~
127.0.0.1 - - [10/Aug/2023:15:45:19 +0000] "GET /ping HTTP/1.1" 301 162 "-" "curl/7.81.0"
total 412
drwxrwxr-x 15 1000 1000   4096 Aug 10 15:26 .
drwxr-xr-x  1 root root   4096 Aug  8 08:16 ..
-rw-rw-r--  1 1000 1000    258 Jun 30 15:18 .editorconfig
-rw-rw-r--  1 1000 1000   1197 Aug 10 15:07 .env
-rw-rw-r--  1 1000 1000   1099 Aug 10 15:07 .env.example
drwxrwxr-x  7 1000 1000   4096 Aug 10 14:52 .git
-rw-rw-r--  1 1000 1000    186 Jun 30 15:18 .gitattributes
-rw-rw-r--  1 1000 1000    255 Aug 10 14:38 .gitignore
-rw-rw-r--  1 1000 1000   4158 Jun 30 15:18 README.md
drwxrwxr-x  3 1000  998   4096 Aug 10 15:12 _volumes
drwxrwxr-x  7 1000 1000   4096 Jun 30 15:18 app
-rwxr-xr-x  1 1000 1000   1686 Jun 30 15:18 artisan
drwxrwxr-x  3 1000 1000   4096 Jun 30 15:18 bootstrap
-rw-rw-r--  1 1000 1000   1919 Aug 10 14:32 composer.json
-rw-rw-r--  1 1000 1000 294977 Aug 10 14:32 composer.lock
drwxrwxr-x  2 1000 1000   4096 Jun 30 15:18 config
drwxrwxr-x  5 1000 1000   4096 Jun 30 15:18 database
drwxrwxr-x  2 1000  998   4096 Aug 10 15:27 docker
-rw-rw-r--  1 1000 1000   1080 Aug 10 15:14 docker-compose.dev.yml
-rw-rw-r--  1 1000 1000    279 Aug 10 15:30 docker-compose.yml
-rw-rw-r--  1 1000 1000    248 Jun 30 15:18 package.json
-rw-rw-r--  1 1000 1000   1084 Jun 30 15:18 phpunit.xml
drwxrwxr-x  2 1000 1000   4096 Jun 30 15:18 public
drwxrwxr-x  5 1000 1000   4096 Jun 30 15:18 resources
drwxrwxr-x  2 1000 1000   4096 Jun 30 15:18 routes
drwxrwxr-x  5 1000 1000   4096 Jun 30 15:18 storage
drwxrwxr-x  4 1000 1000   4096 Jun 30 15:18 tests
drwxrwxr-x 40 1000 1000   4096 Aug 10 14:32 vendor
-rw-rw-r--  1 1000 1000    263 Jun 30 15:18 vite.config.js
~~~

Which is interesting because I expected the user and group ids to be 9999:9999.

**Comment by elcapo** [comment]
> Which is interesting because I expected the user and group ids to be 9999:9999.

For additional context, I tried to run the project changing the 9999 user and group ids to 1000 (the default values) and the permission denied error disappeared. Briefly, I replaced 9999 with 1000 in:

- https://github.com/serversideup/docker-php/blob/290e14f973572ef1ad0a25acb43a88d1dfc19d1e/src/cli/Dockerfile#L30 
- https://github.com/serversideup/docker-php/blob/290e14f973572ef1ad0a25acb43a88d1dfc19d1e/src/cli/etc/s6-overlay/scripts/runas-user#L4
- https://github.com/serversideup/docker-php/blob/290e14f973572ef1ad0a25acb43a88d1dfc19d1e/docs/content/docs/6.reference/1.environment-variable-specification.md?plain=1#L15

Actually, if I now run a `ls -la` command of the webroot, now I get:

~~~
127.0.0.1 - - [10/Aug/2023:16:22:56 +0000] "GET /ping HTTP/1.1" 301 162 "-" "curl/7.81.0"
total 412
drwxrwxr-x 15 webuser webgroup   4096 Aug 10 15:26 .
drwxr-xr-x  1 root    root       4096 Aug 10 16:10 ..
-rw-rw-r--  1 webuser webgroup    258 Jun 30 15:18 .editorconfig
-rw-rw-r--  1 webuser webgroup   1197 Aug 10 15:07 .env
-rw-rw-r--  1 webuser webgroup   1099 Aug 10 15:07 .env.example
drwxrwxr-x  7 webuser webgroup   4096 Aug 10 14:52 .git
-rw-rw-r--  1 webuser webgroup    186 Jun 30 15:18 .gitattributes
-rw-rw-r--  1 webuser webgroup    255 Aug 10 14:38 .gitignore
-rw-rw-r--  1 webuser webgroup   4158 Jun 30 15:18 README.md
drwxrwxr-x  3 webuser      998   4096 Aug 10 15:12 _volumes
drwxrwxr-x  7 webuser webgroup   4096 Jun 30 15:18 app
-rwxr-xr-x  1 webuser webgroup   1686 Jun 30 15:18 artisan
drwxrwxr-x  3 webuser webgroup   4096 Jun 30 15:18 bootstrap
-rw-rw-r--  1 webuser webgroup   1919 Aug 10 14:32 composer.json
-rw-rw-r--  1 webuser webgroup 294977 Aug 10 14:32 composer.lock
drwxrwxr-x  2 webuser webgroup   4096 Jun 30 15:18 config
drwxrwxr-x  5 webuser webgroup   4096 Jun 30 15:18 database
drwxrwxr-x  3 webuser      998   4096 Aug 10 15:53 docker
-rw-rw-r--  1 webuser webgroup   1080 Aug 10 15:14 docker-compose.dev.yml
-rw-rw-r--  1 webuser webgroup    212 Aug 10 16:17 docker-compose.yml
-rw-rw-r--  1 webuser webgroup    248 Jun 30 15:18 package.json
-rw-rw-r--  1 webuser webgroup   1084 Jun 30 15:18 phpunit.xml
drwxrwxr-x  2 webuser webgroup   4096 Aug 10 16:17 public
drwxrwxr-x  5 webuser webgroup   4096 Jun 30 15:18 resources
drwxrwxr-x  2 webuser webgroup   4096 Jun 30 15:18 routes
drwxrwxr-x  5 webuser webgroup   4096 Jun 30 15:18 storage
drwxrwxr-x  4 webuser webgroup   4096 Jun 30 15:18 tests
drwxrwxr-x 40 webuser webgroup   4096 Aug 10 14:32 vendor
-rw-rw-r--  1 webuser webgroup    263 Jun 30 15:18 vite.config.js
~~~

**Comment by sneycampos** [comment]
I am not expert on how the user 9999 is being created but i think we are on the way :)

**Comment by ChrisToxz** [comment]
Till there is a solid fix, I would just suggest to add the following environment variables, to let the image run on a different PUID & PGID.

docker-compose example:
```yaml
environment:
      PUID: 1000
      PGID: 1000

```

**Comment by jaydrogers** [comment]
### Just a note
If you're running this locally, the owner and group IDs will differ based on the host machine.

**If you're using a Linux host to run the docker container**
- You'll probably get WYSIWYG (meaning a container running as the default `9999` will create the file as `9999` on the host)

**If you're using a Windows or macOS to run the container -- it gets complicated**
- Since Docker is not native to these operating systems, Docker Desktop creates a network file share to mount the files to the container, changing the UID & GIDs 

Just another level of complexity I am trying to work through 🙃

**Comment by zogot** [comment]
For me, when I had an issue trying to create a Docker Image that wasn't using volumes my RUN commands were running as root. 

This is my Dockerfile

```
ARG PHP_VERSION='8.2'

# ================
# Base Stage
# ================
FROM serversideup/php:${PHP_VERSION}-fpm-nginx as base
ENV AUTORUN_ENABLED=false
ENV SSL_MODE=off

# ================
# Production Stage
# ================
FROM base as production

ENV APP_ENV=production
ENV APP_DEBUG=false

# Required Modules
RUN apt-get update && \
    apt-get install -y php${PHP_VERSION}-pgsql && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* /usr/share/doc/*

USER $PUID:$PGID

# Copy contents.
# - To ignore files or folders, use .dockerignore
COPY --chown=$PUID:$PGID . .

RUN composer install --optimize-autoloader --no-dev --no-interaction --no-progress --ansi

# artisan commands
RUN php ./artisan event:cache && \
    php ./artisan route:cache && \
    php ./artisan view:cache

USER root:root
```

Note the setting of USER to $PUID:$PGID and then back to root:root for fpm to work correctly.

**Comment by jaydrogers** [comment]
### Quick Question
Can everyone on this thread post their host operating system? I assume it's all Linux?

I am working on resolving this issue. I finally have a Linux development machine to do some testing and reproduction with this issue to finally get it resolved.

**Comment by ChrisToxz** [comment]
Linux, but got team members using WSl & Mac.
Workarounds on Linux works fine on WSl, but Mac seems to be more annoying.. At least development wise with files mounted rather than files inside the image.

**Comment by jaydrogers** [comment]
Hmm... we use macOS exclusively and never run into a problem. Usually with the file mount on macOS, it actually helps prevent the issues from happening.

Do you have steps to reproduce specifically for your Mac? I can run it on my machine and see if I can replicate.

**Comment by gertvdv** [comment]
Experiencing the issue on Ubuntu 22.04 LTS, fresh install.

Note, I'm not seeing this problem with Laravel sail, maybe we can find some inspiration there?

**Comment by elcapo** [comment]
> Can everyone on this thread post their host operating system? I assume it's all Linux?

Another Ubuntu 22.04 LTS here.

**Comment by flemming-petersen** [comment]
Ubuntu 22.04. This error will occur in mounted path like
```
volumes:
  - ./storage:/var/www/html/storage
```
as well in docker volumes
```
volumes:
  - storage:/var/www/html/storage
```

**Comment by jaydrogers** [comment]
I think I have an update on how to fix this. Unfortunately this will require a lot of refactoring.

Keep an eye on this PR. More to come soon 🚀

Any feedback, please let me know 👍

https://github.com/serversideup/docker-php/pull/205

**Comment by XaiMeS** [comment]
Hi,
I have a problem, probably related to this error.
The problem occurs on Linux (fedora) and on Windows (WSL2).

1. when I use my local machine to create a Laravel project (`composer create-project laravel/laravel example-app`) and run docker compose up the following error pops up:
![image](https://github.com/serversideup/docker-php/assets/36962784/94733ce3-059b-434e-844b-6b150027960b)

2. when I use the container to create the Laravel project (`docker run --rm -v $(pwd):/var/www/html serversideup/php:beta-8.3-fpm-nginx-alpine composer create-project laravel/laravel example-app`) then running docker compose up afterwards displays the laravel welcome page, but I cannot edit any files on my local machine.

The problem is probably related to permissions, but because of this, I am not able to use these containers for development.

In older containers, this solution helped: https://github.com/serversideup/docker-php/pull/172#issuecomment-1512779043

But now these variables are no longer available. 

Is there a solution to this problem?


This is my docker compose, it's very simple:
```
version: '3'
services:
    php:
        image: serversideup/php:beta-8.3-fpm-nginx-alpine
        ports:
            - 80:80
        volumes:
            - .:/var/www/html
```

**Comment by sneycampos** [comment]
> Hi, I have a problem, probably related to this error. The problem occurs on Linux (fedora) and on Windows (WSL2).

Probably these environment variables will be added soon, just noticed a note in the v3 docs:

![image](https://github.com/serversideup/docker-php/assets/585263/a1defb81-cd2f-4498-a8bb-e13e42adaa81)

**Comment by XaiMeS** [comment]
Yes, I hope they will be added.

But I wonder how others are using these containers for develpoment? It doesn't work on Windows or Linux. Does it work seamlessly on MacOs? Does everyone only use these containers production-wise? 😃

**Comment by tonysm** [comment]
@XaiMeS I'm using it on Linux for local dev, but I had to set the `PUID` and `PGID` both to `1000` (my host machine's user ID) for it to work. For instance, this is my local `docker-compose.yml`:

```yml
services:
    laravel.test:
        image: serversideup/php:8.2-fpm-nginx
        extra_hosts:
            - 'host.docker.internal:host-gateway'
        ports:
            - '${APP_PORT:-80}:80'
        environment:
            LARAVEL_SAIL: 1
            SSL_MODE: 'off'
            PUID: '${UID:-1000}'
            PGID: '${GID:-1000}'
        volumes:
            - '.:/var/www/html'
        networks:
            - sail
        depends_on:
            - mysql
// ...
```

**Comment by sneycampos** [comment]
> @XaiMeS I'm using it on Linux for local dev, but I had to set the `PUID` and `PGID` both to `1000` (my host machine's user ID) for it to work. For instance, this is my local `docker-compose.yml`:
> 

These env vars works in v2, not in v3 beta

**Comment by tonysm** [comment]
Ah, missed that y'all trying to use the v3 beta. Whoops, sorry.

**Comment by jaydrogers** [comment]
I'll have an update for you guys soon on this in the V3 beta.

@tonysm's [comment](https://github.com/serversideup/docker-php/issues/179#issuecomment-1845519090) is heading in the right direction -- but it can have negative effects depending on what variation you're using.

Unfortunately a lot of this comes down to how Docker & PHP work, so I will be providing improved documentation on how to address permission issues.

I have an idea with [Spin](https://github.com/serversideup/spin/pull/24) on how to make this really easy and automated, but I will need more time to explore this solution.

**Comment by jaydrogers** [comment]
## 👉 Important Update

Hey all,

I just wanted to give you another update that I confirmed I have a working and stable solution with  [Spin](https://github.com/serversideup/spin/pull/24).

I feel this is the best balance that I know of "The Docker Way" and how PHP works.

In short, the best way so solve this is use a build argument to change the user ID:

```Dockerfile
# Learn more about the Server Side Up PHP Docker Images at:
# https://serversideup.net/open-source/docker-php/

FROM serversideup/php:beta-8.3-fpm-nginx as base

FROM base as development

# Fix permission issues in development by setting the "www-data"
# user to the same user and group that is running docker.
ARG USER_ID
ARG GROUP_ID
RUN usermod -u $USER_ID www-data && \
    groupmod -g $GROUP_ID www-data

FROM base as deploy
COPY --chown=www-data:www-data . /var/www/html
```

You can do this in Docker Compose yourself if you wanted to.

```yaml
version: '3.8'
services:

  php:
    build:
      target: development
      args:
        USER_ID: ${SPIN_USER_ID}
        GROUP_ID: ${SPIN_GROUP_ID}
    volumes:
      - .:/var/www/html/
    networks:
      - development
```

### How this works
The above examples will change the `www-data` user when the build target is set to `development`.

In Spin, I automate ALL of this for you:
1. I grab the users user id and group id
2. A `spin up` will automatically grab the development build target
3. The `www-data` user id will be changed to match the user id of the running user

**👉 All of this above prevents any PHP commands from running as a differing user ID**

Best of all, I got this to work on macOS, Linux, and Windows (with WSL2). 🥳

### Next Steps
Stay tuned as I will likely create content around this issue. This is one of those "gotchas" of using Docker + PHP in a production use case.

I will post on here once I have more documentation/videos ready on this 👍

**Comment by sneycampos** [comment]
I like this way and is exactly the way i'm doing right now. Using the env var for id with 1000 as default

**Comment by jaydrogers** [comment]
### Related note
Feel free to chime in on this issue: https://github.com/serversideup/docker-php/issues/253

Right now `www-data` has two separate ids based off if you're using Alpine or Debian in the beta.

I don't like that.

I am thinking of creating a `webuser` and `webgroup` of the ID `9999` like I did the last time. So far that **is not** in the beta yet.

**Comment by jaydrogers** [comment]
# ⚡️ New Feature Added

Hey all,

I just wanted to explain my latest commit: https://github.com/serversideup/docker-php/commit/21156565900949c7a776d93c20beada7ed99bfbf

### What's new
> [!NOTE]  
> This is feature is available in the **v3 beta** images only.

I created a `docker-php-serversideup-set-id` command that allows you to set the permissions easily.

**Usage:**
```bash
docker-php-serversideup-set-id [username] [uid] [gid]
```

### What this does
This takes the username you want to change the UID and GID for. It also checks to make sure other users in the container are not taken. If they are, the script moves them to a new id (hopefully that's safe 😅).

### Why this was created
The intention of this script was to make it easy for a sysadmin to work with a decentralized workforce with many different types of machines and configurations.

It allows a sysadmin to set these files to fix the permission issues when developers are working with their files:

**Dockerfile:**
```dockerfile
# Learn more about the Server Side Up PHP Docker Images at:
# https://serversideup.net/open-source/docker-php/

FROM serversideup/php:8.3.0-fpm-nginx-bookworm as base

# Fix permission issues in development by setting the "www-data"
# user to the same user and group that is running docker.
FROM base as development
ARG USER_ID
ARG GROUP_ID
RUN docker-php-serversideup-set-id www-data ${USER_ID} ${GROUP_ID}

FROM base as deploy
COPY --chown=www-data:www-data . /var/www/html
```

**docker-compose.yml:**
```yaml
version: '3.8'
services:

  php:
    build:
      target: development
      args:
        USER_ID: ${SPIN_USER_ID}
        GROUP_ID: ${SPIN_GROUP_ID}
    volumes:
      - .:/var/www/html/
```

### Long story short
This wasn't really a bug with the Docker images, this was more of a user experience thing due to how PHP, Docker, and File Permissions work.

I created this feature and will provide templates to minimize confusion.

### Next Steps
Let me know if you like this solution and I can close this issue.

Thanks for all your feedback to help get this resolved!

**Comment by syamsoul** [comment]
this error also happened to me when i try to mount volume in kubernetes..

but if i not mount volume, it can start the pod..

![image](https://github.com/serversideup/docker-php/assets/15118790/f0b7537c-df77-41c6-a197-1a3a177168f8)

**Comment by jaydrogers** [comment]
> this error also happened to me when i try to mount volume in kubernetes..
> 
> but if i not mount volume, it can start the pod..
> 
> ![image](https://private-user-images.githubusercontent.com/15118790/294136476-f0b7537c-df77-41c6-a197-1a3a177168f8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQzODkzNzksIm5iZiI6MTcwNDM4OTA3OSwicGF0aCI6Ii8xNTExODc5MC8yOTQxMzY0NzYtZjBiNzUzN2MtZGY3Ny00MWM2LWExOTctMWEzYTE3NzE2OGY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA0VDE3MjQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5ZjFmMjA1MzgyOTcxYTQxZDVkYjVkZDJiZmM5ZWYyZjM4N2VkYjY0MjdiMTI0NTMwNWZkZDFkMGY1OWNhYjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2ODHTbKceLp_NsP40M_0ohZnjbcfU8WCB_M50gWwoU0)

@syamsoul: Try the beta and consider using a different variation than `fpm-nginx` (like `unit`): https://github.com/serversideup/docker-php/discussions/254

**Comment by Quadrubo** [comment]
Hey,
trying the fix right now but I seem to be missing something.

```Dockerfile
FROM serversideup/php:beta-8.2-fpm-nginx as base

FROM base as development

# Fix permission issues in development by setting the "www-data"
# user to the same user and group that is running docker.
ARG USER_ID
ARG GROUP_ID
RUN docker-php-serversideup-set-id www-data ${USER_ID} ${GROUP_ID}

FROM base as deploy

COPY --chown=www-data:www-data . /var/www/html
```

```yml
version: '3.8'
services:
  php:
    build:
      context: .
      dockerfile: Dockerfile
      args:
        USER_ID: 1000
        GROUP_ID: 1000
    ports:
      - 80:80
      - 443:443
    volumes:
      - .:/var/www/html
```

Whenever I start the container I get feedback that the docker user is root and the ids are 0. However I started the container just with docker compose up. My user is in the docker group. I am not executing as root.

```
test-php-1  | Docker user:   root
test-php-1  | Docker uid:    0
test-php-1  | Docker gid:    0
```

**Comment by chikenare** [comment]
If anyone is having permissions issues and using Docker desktop, try uninstalling and using only Docker Engine.
For days I was struggling and in the end it turned out to be Docker Desktop.

**Comment by rburgst** [comment]
@jaydrogers I am trying to run the docker container as non-root as you have outlined in https://github.com/serversideup/docker-php/issues/179#issuecomment-1871385674
However, none of the startup scripts will work in that scenario as they require creating symlinks etc.
This is my `docker-compose.yml`

```yaml
services:
  review:
    image: my/app
    build:
      context: .
      dockerfile: Dockerfile-review
    ports:
      # laravel
      - '8001:81'
    environment:
      PORT: 81
      PUID: '${UID:-1000}'
      PGID: '${GID:-1000}'
    user: 'www-data:1000'
    networks:
      - app-network
 ```
 I am trying to get such a container running on heroku and heroku wont allow you to run as root.

**Comment by jaydrogers** [comment]
@rburgst: You'll need to remove the `user` key. Unfortunately Apache will require `root` to initialize, then it drops permissions to run as `www-data` in the default config.

This is something I plan to improve documentation on, but it's also a limitation with some of these older servers like Apache.

I hope that I can find a true non-privileged execution of PHP with FrankenPHP (https://github.com/serversideup/docker-php/pull/283)

**Comment by rburgst** [comment]
> @rburgst: You'll need to remove the `user` key. Unfortunately Apache will require `root` to initialize, then it drops permissions to run as `www-data` in the default config.
> 
As I am trying to run in heroku this is nothing I have under my control. 
In the meantime I will try with nginx and frankenphp.

**Comment by jaydrogers** [comment]
# 💪 Big Update
So here is a recap of the problem. It all comes down to `root` is required to run on **privileged** ports (80 + 443). We then drop to `www-data` after, which then causes a mess because any Docker `CMD` passed will default to `root` (not `www-data) -- causing this error.

![image](https://github.com/serversideup/docker-php/assets/3174134/21ad66e7-6549-4991-980e-4df8ac77b7b3)

# Proposed Solution
- I'm thinking of defaulting the `fpm-apache` and `fpm-nginx` images to be **UNPRIVILEGED** by default
- This will dramatically improve container security
- This will also reduce the complexities and required workarounds with `su`, `sudo`, etc

# ⚠️ Breaking change
> [!CAUTION]
> This will require users of the `beta-*-fpm-nginx` and `beta-*-fpm-apache` images to make updates

# 👉 What would be required to fix
- By default, our NGINX and Apache images would listen on 8080 and 8443
- If you want port 80 + 443 on your host to go directly to the container, you would need to configure `-p 80:8080 -p 443:8443`

Would love to hear everyone's thoughts 😃

**Comment by AlejandroAkbal** [comment]
Works for me, would like to see it implemented

**Comment by sneycampos** [comment]
There's only one answer and the answer is YES

**Comment by rburgst** [comment]
That would be cool, while I a had some success to get franken-php to work, I still have a long way to go to get it to work properly on heroku.

**Comment by jsayer101** [comment]
> # 💪 Big Update
> So here is a recap of the problem. It all comes down to `root` is required to run on **privileged** ports (80 + 443). We then drop to `www-data` after, which then causes a mess because any Docker `CMD` passed will default to `root` (not `www-data`) -- causing the error on this thread.
> 
> ![image](https://private-user-images.githubusercontent.com/3174134/322133821-21ad66e7-6549-4991-980e-4df8ac77b7b3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMxMTUzNjEsIm5iZiI6MTcxMzExNTA2MSwicGF0aCI6Ii8zMTc0MTM0LzMyMjEzMzgyMS0yMWFkNjZlNy02NTQ5LTQ5OTEtOTgwZS00ZGY4YWM3N2I3YjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTRUMTcxNzQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTA0YTAwNjVlY2Q1ZTc3YmQ0M2ExMjhkYWQxZTQ1NjhhYjc0M2IzZTQ1MTdlZjY4Y2E4NzYyMDJkZTIwYzY4NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.mV24uNIh6wyipW1iLiINf8p1jYZrmjyHw4vbNhPEqvc)
> 
> # 🦄 Proposed Solution
> * I'm thinking of defaulting the `fpm-apache` and `fpm-nginx` images to be **UNPRIVILEGED** by default
> * This will dramatically improve container security
> * This will also reduce the complexities and required workarounds with `su`, `sudo`, etc
> 
> # ⚠️ Breaking change
> Caution
> 
> This will require users of the `beta-*-fpm-nginx` and `beta-*-fpm-apache` images to make updates
> 
> # 👉 What would be required to fix
> * By default, our NGINX and Apache images would listen on 8080 and 8443
> * If you want port 80 + 443 on your host to go directly to the container, you would need to configure `-p 80:8080 -p 443:8443`
> 
> # 🗣️ Related discussions
> * https://twitter.com/tobias_petry/status/1778816006409850997
> * https://twitter.com/jaydrogers/status/1778823124789637216
> 
> Would love to hear everyone's thoughts 😃

Would love to see that merged asap 😄

**Comment by jaydrogers** [comment]
# 💪 Another Big Update
We have a PR ready for testing that will dramatically improve container security and developer experience:
- https://github.com/serversideup/docker-php/pull/311

# 🙏 Please help test this
Everything is documented on the PR on how to test:
- We have pre-built images for you to test (see the PR for more details)
- We also dramatically improved the documentation with the containers moving to be unprivileged by default (get the preview URL from the PR)

We will merge this big change soon if we get good feedback 😅🚀

[Check the PR for more details →](https://github.com/serversideup/docker-php/pull/311)

**Comment by jaydrogers** [comment]
# A fix has been released 🥳
These changes are now live in our `beta` images! Learn more about the release here:
- https://github.com/serversideup/docker-php/releases/tag/v3.0.0-beta4

Monitor this CI/CD job for the exact moment these changes will be live (takes about an hour to build):
- https://github.com/serversideup/docker-php/actions/runs/8788452858

**Comment by chikenare** [comment]
It worked fine for me until a few minutes ago when I did a --build, now I get these errors

Password: su: Authentication failure
Unit is running unprivileged, then it cannot use arbitrary user and group
![Screenshot from 2024-04-25 19-37-34](https://github.com/serversideup/docker-php/assets/25045048/c144357d-9300-4a1d-af58-c991641bb3f4)
